Source/Core/InputCommon/GCAdapter.cpp

```diff
@@ -55,24 +56,30 @@
 namespace GCAdapter
 {
 #if GCADAPTER_USE_LIBUSB_IMPLEMENTATION
+
+constexpr unsigned int USB_TIMEOUT_MS = 16;
+
 static bool CheckDeviceAccess(libusb_device* device);
 static void AddGCAdapter(libusb_device* device);
 static void ResetRumbleLockNeeded();
 #endif
+
 static void Reset();
 static void Setup();
-static void Read();
-static void Write();
+static void ProcessInputPayload(const u8* data, std::size_t size);
+static void ReadThreadFunc();
+static void WriteThreadFunc();
 
 #if GCADAPTER_USE_LIBUSB_IMPLEMENTATION
-enum
+enum class AdapterStatus
 {
-  NO_ADAPTER_DETECTED = 0,
-  ADAPTER_DETECTED = 1,
+  NotDetected,
+  Detected,
+  Error,
 };
 
-// Current adapter status: detected/not detected/in error (libusb error codes are negative)
-static std::atomic<int> s_status = NO_ADAPTER_DETECTED;
+static std::atomic<AdapterStatus> s_status = AdapterStatus::NotDetected;
+static std::atomic<libusb_error> s_adapter_error = LIBUSB_SUCCESS;
```

Comment:
Another thought here - does it make sense to have 2 atomic variables like this? I don't think it'd be atomic anymore when done this way. Though `IsDetected` also reads (and previously read) `s_status` multiple times. (You were the one who initially made `s_status` atomic in #8933, for what it's worth; also see #10540 and #10604.)

Other than that, looks good to me.

Reply:
They need to be at least atomic because they are read/written from different threads.
Even though their combined usage is not atomic it should be fine since it just generates UI error text.
A race should ultimately resolve to the correct UI message after the next adapter callback.